src/Compilers/Core/Portable/SourceFileResolver.cs

```diff
@@ -39,21 +48,65 @@ public SourceFileResolver(ImmutableArray<string> searchPaths, string baseDirecto
 
             _baseDirectory = baseDirectory;
             _searchPaths = searchPaths;
-        }
+            _pathMap = pathMap;
 
-        public string BaseDirectory
-        {
-            get { return _baseDirectory; }
+            // the keys in pathMap should not end with a path separator
+            if (!pathMap.IsDefaultOrEmpty)
+            {
+                foreach (var kv in pathMap)
+                {
+                    var key = kv.Key;
+                    if (key == null || key.Length == 0)
+                    {
+                        throw new ArgumentException(CodeAnalysisResources.EmptyKeyInPathMap, nameof(pathMap));
+                    }
+
+                    if (IsPathSeparator(key[key.Length - 1]))
+                    {
+                        throw new ArgumentException(CodeAnalysisResources.KeyInPathMapEndsWithSeparator, nameof(pathMap));
+                    }
+                }
+            }
         }
 
-        public ImmutableArray<string> SearchPaths
+        public string BaseDirectory => _baseDirectory;
+
+        public ImmutableArray<string> SearchPaths => _searchPaths;
+
+        public ImmutableArray<KeyValuePair<string, string>> PathMap => _pathMap;
+
+        public override string NormalizePath(string path, string baseFilePath)
         {
-            get { return _searchPaths; }
+            string normalizedPath = FileUtilities.NormalizeRelativePath(path, baseFilePath, _baseDirectory);
+            return (normalizedPath == null || _pathMap.IsDefaultOrEmpty) ? normalizedPath : NormalizePathPrefix(normalizedPath, _pathMap);
         }
 
-        public override string NormalizePath(string path, string baseFilePath)
+        private string NormalizePathPrefix(string normalizedPath, ImmutableArray<KeyValuePair<string, string>> pathMap)
```

Comment:
The method isn't static. Is there a reason to pass pathMap as an argument when it is an instance field?